File: lint-arwaky/archive/contract/plugin_manager_port.py
"""plugin_manager_port — Interface for plugin discovery and management."""

from abc import ABC, abstractmethod
from ..taxonomy import AdapterName, PluginGroup, AdapterMetadataList, AdapterClassMap, PluginError


class IPluginManagerPort(ABC):
    """Port for discovering and managing third-party adapters via entry points."""

    @abstractmethod
    def discover_plugins(self, group: PluginGroup) -> AdapterClassMap | PluginError:
        """Discover custom adapters via entry points and register them."""
        ...

    @abstractmethod
    def list_custom_adapters(self) -> AdapterMetadataList:
        """Return a list of all registered custom adapters with metadata."""
        ...

    @abstractmethod
    def register_custom_adapter(self, name: AdapterName, adapter_class: type) -> PluginError | None:
        """Manually register a custom adapter class."""
        ...

File: lint-arwaky/src-rust/contract/plugin_manager_port.rs
use super::*;

pub trait IPluginManagerPort: Send + Sync {
    fn discover_plugins(&self, group: &str) -> Result<Vec<(String, String)>, PluginError>;
    fn list_custom_adapters(&self) -> Vec<AdapterMetadata>;
    fn register_custom_adapter(&self, name: &AdapterName, class_path: &str) -> Result<(), PluginError>;
}
